### ISSUE TITLE
Failed building wheel for pytinyxml2

### ISSUE THREAD

**Issue by ivan-enzhaev** [jiegec/pytinyxml2#1]
- Python 3.8
- pytinyxml2 9.0.0
- Window 10

Failed to install pytinyxml2:

```
>pip install pytinyxml2
Collecting pytinyxml2
  Using cached pytinyxml2-9.0.0.tar.gz (44 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: pytinyxml2
  Building wheel for pytinyxml2 (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pytinyxml2 (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [19 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-38
      copying pytinyxml2.py -> build\lib.win-amd64-cpython-38
      running egg_info
      writing pytinyxml2.egg-info\PKG-INFO
      writing dependency_links to pytinyxml2.egg-info\dependency_links.txt
      writing top-level names to pytinyxml2.egg-info\top_level.txt
      reading manifest file 'pytinyxml2.egg-info\SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      adding license file 'LICENSE'
      writing manifest file 'pytinyxml2.egg-info\SOURCES.txt'
      running build_ext
      building '_pytinyxml2' extension
      swigging pytinyxml2.i to pytinyxml2_wrap.cpp
      swig.exe -python -c++ -o pytinyxml2_wrap.cpp pytinyxml2.i
      error: command 'swig.exe' failed: None
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pytinyxml2
Failed to build pytinyxml2
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (pytinyxml2)
```

**Comment by ivan-enzhaev** [comment]
> error: command 'swig.exe' failed: None

I have installed `swig` using this tutorial: https://www.dev2qa.com/how-to-install-swig-on-macos-linux-and-windows/

```
>swig -version

SWIG Version 3.0.12

Compiled with i686-w64-mingw32-g++ [i686-w64-mingw32]

Configured options: +pcre

Please see http://www.swig.org for reporting bugs and further information
```

```
>pip install pytinyxml2
Collecting pytinyxml2
  Using cached pytinyxml2-9.0.0.tar.gz (44 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: pytinyxml2
  Building wheel for pytinyxml2 (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pytinyxml2 (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [29 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-38
      copying pytinyxml2.py -> build\lib.win-amd64-cpython-38
      running egg_info
      writing pytinyxml2.egg-info\PKG-INFO
      writing dependency_links to pytinyxml2.egg-info\dependency_links.txt
      writing top-level names to pytinyxml2.egg-info\top_level.txt
      reading manifest file 'pytinyxml2.egg-info\SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      adding license file 'LICENSE'
      writing manifest file 'pytinyxml2.egg-info\SOURCES.txt'
      running build_ext
      building '_pytinyxml2' extension
      swigging pytinyxml2.i to pytinyxml2_wrap.cpp
      swig.exe -python -c++ -o pytinyxml2_wrap.cpp pytinyxml2.i
      tinyxml2.h(260) : Warning 389: operator[] ignored (consider using %extend)
      tinyxml2.h(265) : Warning 389: operator[] ignored (consider using %extend)
      tinyxml2.h(2068) : Warning 362: operator= ignored
      tinyxml2.h(2146) : Warning 362: operator= ignored
      tinyxml2.h(568) : Warning 509: Overloaded method tinyxml2::XMLUtil::SkipWhiteSpace(char *const,int *) effectively ignored,
      tinyxml2.h(565) : Warning 509: as it is shadowed by tinyxml2::XMLUtil::SkipWhiteSpace(char const *,int *).
      tinyxml2.h(2060) : Warning 509: Overloaded method tinyxml2::XMLHandle::XMLHandle(tinyxml2::XMLNode &) effectively ignored,
      tinyxml2.h(2057) : Warning 509: as it is shadowed by tinyxml2::XMLHandle::XMLHandle(tinyxml2::XMLNode *).
      tinyxml2.h(2139) : Warning 509: Overloaded method tinyxml2::XMLConstHandle::XMLConstHandle(tinyxml2::XMLNode const &) effectively ignored,
      tinyxml2.h(2137) : Warning 509: as it is shadowed by tinyxml2::XMLConstHandle::XMLConstHandle(tinyxml2::XMLNode const *).
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pytinyxml2
Failed to build pytinyxml2
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (pytinyxml2)
```

**Comment by ivan-enzhaev** [comment]
> error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/

It requires "Microsoft C++ Build Tools" but it is too much space. It's 6.75 GB!

![image](https://github.com/user-attachments/assets/959cf5ff-82a4-4212-b090-8495e5e52cb8)

![image](https://github.com/user-attachments/assets/e3cf9aa1-fbc4-4147-a9d4-ca61e7d70f12)

**Comment by jiegec** [comment]
I have only tested it on Linux.

**Comment by ivan-enzhaev** [comment]
Hopefully in the future it will be possible to install pytinyxml2 without installing Visual Studio:
> error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/

**Comment by jiegec** [comment]
> Hopefully in the future it will be possible to install pytinyxml2 without installing Visual Studio:
> 
> > error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/

I don't have Windows environment, but if you figured out how to install it without installing Visual Studio, I am glad to add the instructions to README.

**Comment by ivan-enzhaev** [comment]
I have tried to install "Microsoft C++ Build Tools" without installing Visual Studio using this guide and script: https://gist.github.com/mmozeiko/7f3162ec2988e81e56d5c4e22cde9977/

It requires only 1 GB (not 6.5 GB like using Microsoft's official installer). But it didn't help - there is the same error as above. I think Visual Studio must be installed. The `swing` dependency is bad too, it must be deleted. But I don't know how to solve these two problems. I have found a temporary solution to use `pugxml`: https://pypi.org/project/pugixml/ But I don't like it. I use `tinyxml2` in C++ and I want to use the same tools in Python, I mean: `PySDL`, `PyGLM`, `PyOpenGL`, `PyOpenAL` and `pytinyxml2`. I am watching for all activities of the `pytinyxml2`. I am sure that these problems will be solved in the future because `tinyxml2` is excellent.
